sql/plan/call.go

```diff
@@ -0,0 +1,172 @@
+var _ sql.Expressioner = (*Call)(nil)
+
+// NewCall returns a *Call node.
+func NewCall(name string, params []sql.Expression) *Call {
+	return &Call{
+		Name:   name,
+		Params: params,
+	}
+}
+
+// Resolved implements the sql.Node interface.
+func (c *Call) Resolved() bool {
+	for _, param := range c.Params {
+		if !param.Resolved() {
+			return false
+		}
+	}
+	return true
+}
+
+// Schema implements the sql.Node interface.
+func (c *Call) Schema() sql.Schema {
+	return nil
+}
+
+// Children implements the sql.Node interface.
+func (c *Call) Children() []sql.Node {
+	return nil
+}
+
+// WithChildren implements the sql.Node interface.
+func (c *Call) WithChildren(children ...sql.Node) (sql.Node, error) {
+	return NillaryWithChildren(c, children...)
+}
+
+// Expressions implements the sql.Expressioner interface.
+func (c *Call) Expressions() []sql.Expression {
+	return c.Params
+}
+
+// WithExpressions implements the sql.Expressioner interface.
+func (c *Call) WithExpressions(exprs ...sql.Expression) (sql.Node, error) {
+	if len(exprs) != len(c.Params) {
+		return nil, fmt.Errorf("%s: invalid param number, got %d, expected %d", c.Name, len(exprs), len(c.Params))
+	}
+
+	nc := *c
+	nc.Params = exprs
+	return &nc, nil
+}
+
+// WithProcedure returns a new *Call containing the given *CreateProcedure.
+func (c *Call) WithProcedure(proc *CreateProcedure) *Call {
+	nc := *c
+	nc.proc = proc
+	return &nc
+}
+
+// HasProcedure returns whether a *Call has had its procedure set.
+func (c *Call) HasProcedure() bool {
+	return c.proc != nil
+}
+
+// WithParamReference returns a new *Call containing the given *expression.ProcedureParamReference.
+func (c *Call) WithParamReference(pRef *expression.ProcedureParamReference) *Call {
+	nc := *c
+	nc.pRef = pRef
+	return &nc
+}
+
+// String implements the sql.Node interface.
+func (c *Call) String() string {
+	paramStr := ""
+	for i, param := range c.Params {
+		if i > 0 {
+			paramStr += ", "
+		}
+		paramStr += param.String()
+	}
+	return fmt.Sprintf("CALL %s(%s)", c.Name, paramStr)
+}
+
+// RowIter implements the sql.Node interface.
+func (c *Call) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	return &callIter{
+		call: c,
+		ctx:  ctx,
+	}, nil
+}
+
+// callIter is the row iterator for *Call.
+type callIter struct {
+	once sync.Once
+	call *Call
+	ctx  *sql.Context
+}
+
+// Next implements the sql.RowIter interface.
+func (iter *callIter) Next() (sql.Row, error) {
+	run := false
+	iter.once.Do(func() {
+		run = true
+	})
+	if !run {
+		return nil, io.EOF
+	}
+
+	for i, paramExpr := range iter.call.Params {
+		val, err := paramExpr.Eval(iter.ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		paramName := strings.ToLower(iter.call.proc.Params[i].Name)
+		iter.call.pRef.NameToParam[paramName] = val
+	}
+
+	bodyIter, err := iter.call.proc.Body.RowIter(iter.ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var fullRow sql.Row //TODO: figure out something better to do
```

Comment:
Definitely not the right way to go here. Your iter needs to wrap the body iter and return one row for every row it does, until it's at the end. Hard to get the details right.